firedrake/deflation.py

```diff
@@ -0,0 +1,217 @@
+from firedrake.preconditioners.base import SNESBase
+from firedrake import dmhooks
+from firedrake.dmhooks import get_appctx
+from firedrake.petsc import PETSc
+from firedrake import inner, dx
+
+from firedrake_citations import Citations
+
+import weakref
+
+__all__ = ['DeflatedSNES', 'Deflation']
+
+
+class DeflatedSNES(SNESBase):
+    """
+    A SNES that implements deflation, an algorithm for finding
+    multiple solutions.
+
+    It fetches the solutions to deflate and the notion of distance
+    to use from the problem appctx.
+
+    In practice, deflation only requires postprocessing the Newton
+    direction after the linear solve. We use a custom KSP for this
```

Comment:
Why is this done by the KSP? Shouldn't we overload `DeflatedSNES.step()` to deal with the scaling of the Newton direction?

Reply:
If I overload `DeflatedSNES.step`, then I have to manually control the inner SNES to only do one step, duplicate  the termination logic, etc. I think that's more invasive than this way.